src/Microsoft.OData.Core/Evaluation/ODataConventionalEntityMetadataBuilder.cs

```diff
@@ -645,26 +645,29 @@ private Uri GetContainingEntitySetUri(Uri baseUri, ODataUri odataUri)
         {
             ODataPath path = odataUri.Path;
             List<ODataPathSegment> segments = path.ToList();
+            int segmentsCount = segments.Count - 1;
             ODataPathSegment lastSegment = segments.Last();
             while (!(lastSegment is NavigationPropertySegment) && !(lastSegment is OperationSegment))
             {
-                segments.Remove(lastSegment);
-                lastSegment = segments.Last();
+                segmentsCount--;
+                lastSegment = segments[segmentsCount];
             }
 
             // also removed the last navigation property segment
-            segments.Remove(lastSegment);
+            segments = segments.GetRange(0, segmentsCount);
 
             // Remove the unnecessary type cast segment.
             ODataPathSegment nextToLastSegment = segments.Last();
+            int lastIndex = segments.Count - 1;

```

Comment:
@marabooy You could initialize `nextToLastSegment` on this line and forgo use of `.Last()` - unless there's no obvious benefit in doing that, i.e.
```
ODataPathSegment nextToLastSegment = segments[lastIndex];
```
Actually, why is the variable named `nextToLastSegment` yet its seems to me like it's assigned the last segment. Maybe it should be `typecastSegment`